Lib/fontTools/otlLib/optimize.py

```diff
@@ -180,7 +181,42 @@ def cluster_pairs_by_class2_coverage_agglomerative(
     return [*grouped_lines.values(), *ungrouped_lines]
 
 
+# Adapted from https://github.com/fonttools/fonttools/blob/f64f0b42f2d1163b2d85194e0979def539f5dca3/Lib/fontTools/ttLib/tables/otTables.py#L935-L958
+def _getClassRanges(classes: Iterable[Tuple[str]], font: TTFont):
+    getGlyphID = font.getGlyphID
+    items = []
+    for cls, glyphNames in enumerate(classes):
+        for glyphName in glyphNames:
+            items.append((getGlyphID(glyphName), glyphName, cls))
```

Comment:
`getGlyphID` is killing your performance. See this for how to do it faster:

https://github.com/fonttools/fonttools/blob/f64f0b42f2d1163b2d85194e0979def539f5dca3/Lib/fontTools/ttLib/tables/otConverters.py#L334-L352

Also see:
https://github.com/fonttools/fonttools/issues/1536

Maybe now's time to add `TTFont` API for that. And clean up all the ClassDef/Coverage code you referenced.